autogen/oai/client.py

```diff
@@ -141,17 +359,56 @@ def _client(self, config: Dict[str, Any], openai_config: Dict[str, Any]) -> Open
         "gpt-35-turbo" and define model "gpt-3.5-turbo" in the config the function will remove the dot
         from the name and create a client that connects to "gpt-35-turbo" Azure deployment.
         """
+        activated = False
         openai_config = {**openai_config, **{k: v for k, v in config.items() if k in self.openai_kwargs}}
         api_type = config.get("api_type")
-        if api_type is not None and api_type.startswith("azure"):
-            openai_config["azure_deployment"] = openai_config.get("azure_deployment", config.get("model"))
-            if openai_config["azure_deployment"] is not None:
-                openai_config["azure_deployment"] = openai_config["azure_deployment"].replace(".", "")
-            openai_config["azure_endpoint"] = openai_config.get("azure_endpoint", openai_config.pop("base_url", None))
-            client = AzureOpenAI(**openai_config)
+        if api_type is None:
+            self._clients.append(OpenAIClient(OpenAI(**openai_config)))
+            activated = True
         else:
-            client = OpenAI(**openai_config)
-        return client
+            if api_type.startswith("azure"):
+                self._configure_azure_openai(config, openai_config)
+                self._clients.append(OpenAIClient(AzureOpenAI(**openai_config)))
+                activated = True
+            elif api_type.startswith("custom"):
+                # else a config for a custom client is set
+                # skipping until the register_model_client is called with the appropriate class
+                logging.info(
+                    "Detected custom model client in config, skipping registration until register_model_client is called."
+                )
+            else:
+                raise ValueError(
+                    f"api_type {api_type} is not supported, please select one from ['azure', 'custom'], or remove and let it default to openai"
+                )
+        return activated
+
+    def register_model_client(self, model: str, model_client_cls: Client, **kwargs):
+        """Register a model client client.
+
+        Args:
+            model: The model name, as specified in the config list
+            model_client_cls: A custom client class that follows the Client interface
+            **kwargs: The kwargs for the custom client class to be initialized with
+        """
+        for i, config in enumerate(self._config_list):
+            config_model = config.get("config", {}).get("model")
+            activated = config.get("activated")
+
+            if config_model == model:
+                if activated and model_client_cls.__name__ == self._clients[i].__class__.__name__:
```

Comment:
I am not sure if we can just compare with `_clients[i]`? Should we actually scan both client list and config list to make sure this client is not already registered with the target `model`?

Reply:
yes I'll update